In [1]:
#  данные X (темп прироста зарплаты) и Y (темп прироста безработицы )
X = [530.16, 545.98, 552.03, 526.32, 594.75, 557.53, 643.63, 524.82, 447.75, 385.54, 487.23, 632.77]
Y = [7.2, 7.2, 7.1, 7.3, 6.5, 8.5, 8.4, 9.2, 9.8, 11.6, 8.8, 7]

In [5]:
import numpy as np
import statsmodels.api as sm
from statsmodels.stats.diagnostic import het_breuschpagan, het_white, het_goldfeldquandt
from scipy.stats import bartlett

# данные X (темп прироста зарплаты) и Y (темп прироста безработицы )
X = [530.16, 545.98, 552.03, 526.32, 594.75, 557.53, 643.63, 524.82, 447.75, 385.54, 487.23, 632.77]
Y = [7.2, 7.2, 7.1, 7.3, 6.5, 8.5, 8.4, 9.2, 9.8, 11.6, 8.8, 7]

# 1. Получите остатки модели регрессии Y на X
X = sm.add_constant(X)  # Добавляем константу (intercept) к X
model = sm.OLS(Y, X).fit()  # Строим регрессию Y на X
residuals = model.resid  # Получаем остатки
print("Остатки регрессии:")
print(residuals)

# 2. Проведем 3 теста на гомоскедастичность

# Тест Бройша-Пагана
bp_test_stat, bp_p_value, _, _ = het_breuschpagan(residuals, X)
print(f"Тест Бройша-Пагана: статистика = {bp_test_stat}, p-value = {bp_p_value}")

# Тест Уайта
white_test_stat, white_p_value, _, _ = het_white(residuals, X)
print(f"Тест Уайта: статистика = {white_test_stat}, p-value = {white_p_value}")

# Тест Гольдфельда-Квандта
gq_test_stat, gq_p_value, _ = het_goldfeldquandt(residuals, X)
print(f"Тест Гольдфельда-Квандта: статистика = {gq_test_stat}, p-value = {gq_p_value}")

# Тест Барлетта
bartlett_test_stat, bartlett_p_value = bartlett(Y, residuals)
print(f"Тест Барлетта: статистика = {bartlett_test_stat}, p-value = {bartlett_p_value}")



Остатки регрессии:
[-1.1018152  -0.85906707 -0.86623356 -1.06073763 -0.81072155  0.61816054
  1.83931186  0.8162458   0.23365425  1.07908023 -0.16054953  0.27267187]
Тест Бройша-Пагана: статистика = 1.2579711398258442, p-value = 0.26203547577910213
Тест Уайта: статистика = 3.369547255160619, p-value = 0.18548641621261044
Тест Гольдфельда-Квандта: статистика = 1.2319672027586666, p-value = 0.42233353034117327
Тест Барлетта: статистика = 1.8475667156312388, p-value = 0.17406666515042613


Рассмотрим результаты и сделаем выводы:

Тест Бройша-Пагана:

Статистика теста: 1.258
p-значение: 0.262
По результатам теста Бройша-Пагана, нулевая гипотеза о гомоскедастичности не отвергается (p-значение > 0.05). Это означает, что нет значительных доказательств наличия гетероскедастичности в остатках.

Тест Уайта:

Статистика теста: 3.370
p-значение: 0.185
По результатам теста Уайта, также не удается отвергнуть нулевую гипотезу о гомоскедастичности (p-значение > 0.05). Этот тест также не обнаружил статистически значимой гетероскедастичности.

Тест Гольдфельда-Квандта:

Статистика теста: 1.232
p-значение: 0.422
Тест Гольдфельда-Квандта также не позволяет отвергнуть гипотезу о гомоскедастичности (p-значение > 0.05).

Тест Барлетта:

Статистика теста: 1.848
p-значение: 0.174
Результаты теста Барлетта также не подтверждают наличие гетероскедастичности в остатках (p-значение > 0.05).

Исходя из результатов всех четырех тестов, можно сделать вывод, что остатки регрессии не показывают статистически значимой гетероскедастичности. Гипотеза о гомоскедастичности остатков не отвергается на уровне значимости 0.05.

##№2
Тест Бройша-Пагана (Breusch-Pagan Test):

Условия применения: Тест Бройша-Пагана применяется для проверки гомоскедастичности в множественной регрессии. Нулевая гипотеза состоит в том, что остатки регрессии являются гомоскедастичными, то есть дисперсия ошибок постоянна. Если p-значение теста меньше выбранного уровня значимости, то нулевая гипотеза отвергается в пользу альтернативной гипотезы о наличии гетероскедастичности.

Источник: Григорьева, Т. Л. (2013). Эконометрика: Учебное пособие. Москва: Дело.

Тест Уайта (White Test):

Условия применения: Тест Уайта также используется для проверки гомоскедастичности в множественной регрессии. Нулевая гипотеза заключается в том, что остатки регрессии являются гомоскедастичными. Если p-значение теста меньше уровня значимости, то нулевая гипотеза отвергается.

Источник: Демешев, Б. Б., & Малиновская, О. О. (2007). Эконометрика: Учебное пособие. Москва: Издательство "Экономика".

Тест Гольдфельда-Квандта (Goldfeld-Quandt Test):

Условия применения: Тест Гольдфельда-Квандта используется для проверки гомоскедастичности в регрессиях с изменяющейся дисперсией остатков в зависимости от значений независимых переменных. Нулевая гипотеза предполагает гомоскедастичность остатков. Если p-значение меньше уровня значимости, то нулевая гипотеза отвергается.

Источник: Гурьев, С. В. (2016). Эконометрика. Москва: Издательство "Эксмо".

Тест Барлетта (Bartlett Test):

Условия применения: Тест Барлетта обычно используется в анализе дисперсии (ANOVA) для проверки гомоскедастичности. Нулевая гипотеза заключается в том, что дисперсии в различных группах данных равны. Если p-значение теста меньше уровня значимости, то нулевая гипотеза отвергается.

Источник: Ширяев, А. Н. (1980). Вероятность. Москва: Наука